app/controllers/samples_controller.rb

```diff
@@ -721,30 +722,40 @@ def samples_going_public
     render json: samples.to_json(include: [{ project: { only: [:id, :name] } }])
   end
 
+  # The json response here should be precached in PipelineRun.
   def report_info
-    expires_in 30.days
-    @pipeline_run = select_pipeline_run(@sample, params)
-
-    ##################################################
-    ## Duct tape for changing background id dynamically
-    ## TODO(yf): clean the following up.
-    ####################################################
-    if @pipeline_run && (((@pipeline_run.adjusted_remaining_reads.to_i > 0 || @pipeline_run.finalized?) && !@pipeline_run.failed?) || @pipeline_run.report_ready?)
-      background_id = get_background_id(@sample)
-      pipeline_run_id = @pipeline_run.id
+    MetricUtil.put_metric_now("samples.cache.requested", 1)
+
+    pipeline_run = select_pipeline_run(@sample, params[:pipeline_version])
+    if pipeline_run.nil?
+      message = "Pipeline run not found for sample #{@sample.id}"
+      raise ActiveRecord::RecordNotFound, message
     end
+    report_info_params = pipeline_run.report_info_params
 
-    @report_info = external_report_info(pipeline_run_id, background_id, params)
+    # Set background_id to a viewable background
+    params[:background_id] = get_background_id(@sample)
 
-    # Fill lineage details into report info.
-    # @report_info[:taxonomy_details][2] is the array of taxon rows (which are hashes with keys like tax_id, name, NT, etc)
-    @report_info[:taxonomy_details][2] = TaxonLineage.fill_lineage_details(@report_info[:taxonomy_details][2], pipeline_run_id)
+    cache_key = ReportHelper.report_info_cache_key(
+      request.path,
+      params.permit(report_info_params.keys)
+    )
 
-    # Label top-scoring hits for the executive summary
-    @report_info[:topScoringTaxa] = label_top_scoring_taxa!(@report_info[:taxonomy_details][2])
-    @report_info[:contig_taxid_list] = @pipeline_run.get_taxid_list_with_contigs
+    # This allows 304 Not Modified to be returned so that the client can use its
+    # local cache and avoid the large download.
+    httpdate = Time.at(report_info_params[:report_ts]).utc.httpdate
+    response.headers["Last-Modified"] = httpdate
```

Comment:
Based on my understanding, you need to do a little more to get 304 Not Modified to work.
https://blog.appsignal.com/2018/05/01/client-side-caching-in-rails-conditional-get-requests.html

Did you get 304 Not Modified working? If not, maybe defer to a follow-up commit? (and remove the relevant from this one)

Reply:
I did get the 304 working in my browser though I'm not sure if it is the most robust way.